lib/internal/loader/Loader.js

```diff
@@ -34,39 +34,47 @@ class Loader {
     this.base = base;
   }
 
-  async resolve(specifier) {
-    const request = resolveRequestUrl(this.base, specifier);
-    if (request.url.protocol !== 'file:') {
+  setModuleResolver(resolver) {
+    resolve = resolver;
+  }
+
+  async resolve(specifier, parentURLOrString = this.base) {
+    if (typeof parentURLOrString === 'string') {
+      parentURLOrString = new URL(parentURLOrString);
+    }
+    else if (parentURLOrString instanceof URL === false) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'parentURLOrString', 'URL');
+    }
+    const { url, format } = await resolve(specifier, parentURLOrString);
+    if (typeof url === 'string') {
+      url = new URL(url);
+    }
+    else if (!(url instanceof URL)) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'url', 'URL');
+    }
+    if (url.protocol !== 'file:') {
       throw new errors.Error('ERR_INVALID_PROTOCOL',
                              request.url.protocol, 'file:');
     }
-    return request.url;
+    if (!formatProviders.has(format)) {
+      throw new errors.Error('ERR_INVALID_FORMAT', format);
+    }
+    return { url, format };
   }
 
-  async getModuleJob(dependentJob, specifier) {
-    if (!this.moduleMap.has(dependentJob.url)) {
-      throw new errors.Error('ERR_MISSING_MODULE', dependentJob.url);
+  async getModuleJob(specifier, parentURLOrString = this.base) {
+    const { url, format } = await this.resolve(specifier, parentURLOrString);
+    const urlString = `${url}`;
+    let job = this.moduleMap.get(urlString);
```

Comment:
please check using .has

Reply:
This was done as a perf optimization to avoid repeated unnecessary dictionary lookups, but unsure how much it would affect things. Happy to change it.